### PR TITLE
[full-ci] Refactor stores

### PR DESCRIPTION
**Pull request by aduffeck**
This draft PR refactors and streamlines the different caches and stores and the according environment variables to remove duplicated code, confusion and differences in available cache options.
It also tries to distinguish more cleanly between stores (persistence layer used for persisting data and caches) and caches.

Depends on https://github.com/cs3org/reva/pull/3777.

After a discussion with the doc team it was concluded that this is a good point in time to drop support of the old and confusing environment variables and only use the new streamlined versions.

As a result the following variables have been

removed:

```
OCIS_CACHE_STORE_TYPE (use OCIS_CACHE_STORE instead)
OCIS_CACHE_STORE_ADDRESSES (use OCIS_CACHE_STORE_NODES instead)
OCIS_PERSISTENT_STORE_TYPE (use OCIS_PERSISTENT_STORE instead)
EVENTHISTORY_STORE_TYPE (use EVENTHISTORY_STORE instead)
EVENTHISTORY_STORE_ADDRESSES (use EVENTHISTORY_STORE_NODES instead)
GATEWAY_CACHE_STORE_TYPE (use GATEWAY_CACHE_STORE instead)
GATEWAY_CACHE_STORE_ADDRESS (use GATEWAY_CACHE_STORE_NODES instead)
GATEWAY_CACHE_NODES (use GATEWAY_CACHE_STORE_NODES instead)
GRAPH_CACHE_STORE_TYPE (use GRAPH_CACHE_STORE instead)
OCS_CACHE_STORE_TYPE (use OCS_CACHE_STORE instead)
OCS_CACHE_STORE_ADDRESSES (use OCS_CACHE_STORE_NODES instead)
PROXY_OIDC_USERINFO_CACHE_TYPE (use PROXY_OIDC_USERINFO_CACHE_STORE instead)
PROXY_OIDC_USERINFO_CACHE_ADDRESSES (use PROXY_OIDC_USERINFO_CACHE_STORE_NODES instead)
STORAGE_USERS_CACHE_STORE_TYPE (use STORAGE_USERS_CACHE_STORE instead)
STORAGE_USERS_CACHE_STORE_ADDRESS (use STORAGE_USERS_CACHE_STORE_NODES instead)
STORAGE_USERS_CACHE_NODES (use STORAGE_USERS_CACHE_STORE_NODES instead)
USERLOG_STORE_TYPE (use USERLOG_STORE instead)
USERLOG_STORE_ADDRESSES (use USERLOG_STORE_NODES instead)
GATEWAY_CACHE_DATABASE (uses OCIS_CACHE_DATABASE now internally)
PROXY_OIDC_USERINFO_CACHE_DATABASE (uses OCIS_CACHE_DATABASE now internally)
STORAGE_SYSTEM_CACHE_DATABASE (uses OCIS_CACHE_DATABASE now internally)
STORAGE_USERS_CACHE_DATABASE (uses OCIS_CACHE_DATABASE now internally)
```

renamed: 

```
OCIS_CACHE_STORE_TTL -> OCIS_CACHE_TTL
OCIS_CACHE_STORE_SIZE -> OCIS_CACHE_SIZE
EVENTHISTORY_RECORD_EXPIRY -> EVENTHISTORY_STORE_TTL
FRONTEND_OCS_RESOURCE_INFO_CACHE_TYPE -> FRONTEND_OCS_RESOURCE_INFO_CACHE_STORE
GRAPH_CACHE_STORE_TTL -> GRAPH_CACHE_TTL
GRAPH_CACHE_STORE_SIZE -> GRAPH_CACHE_SIZE
OCS_CACHE_STORE_TTL -> OCS_CACHE_TTL
OCS_CACHE_STORE_SIZE -> OCS_CACHE_SIZE
```

added:

```
FRONTEND_OCS_RESOURCE_INFO_CACHE_STORE_NODES
FRONTEND_OCS_RESOURCE_INFO_CACHE_TABLE
FRONTEND_OCS_RESOURCE_INFO_CACHE_TTL
FRONTEND_OCS_RESOURCE_INFO_CACHE_SIZE
GATEWAY_STAT_CACHE_SIZE
GATEWAY_PROVIDER_CACHE_SIZE
GATEWAY_CREATE_HOME_CACHE_SIZE
STORAGE_SYSTEM_CACHE_TTL
STORAGE_SYSTEM_CACHE_SIZE
```

/cc @mmattel

### CODE REVIEW

File: services/frontend/pkg/config/config.go [hunk 1]
```diff
@@ -138,7 +142,7 @@ type ResourceInfoCaches struct {
 
 // RedisDriver holds redis configuration
 type RedisDriver struct {
-	Address  string `yaml:"address" env:"FRONTEND_OCS_RESOURCE_INFO_CACHE_REDIS_ADDR" desc:"A comma separated list of addresses to access the configured store. This has no effect when the 'memory' store is configured. Note that the behaviour how addresses are used is dependent on the library of the configured store."`
+	Address  string `yaml:"address" env:"FRONTEND_OCS_RESOURCE_INFO_CACHE_REDIS_ADDR" desc:"A comma separated list of nodes to access the configured store. This has no effect when 'memory' or 'ocmem' stores are configured. Note that the behaviour how nodes are used is dependent on the library of the configured store."`
```
Comment (mmattel) [inline services/frontend/pkg/config/config.go:145]: `FRONTEND_OCS_RESOURCE_INFO_CACHE_REDIS_ADDR`
`_ADDR` shoud'nt that be `_NODES`

Comment (aduffeck) [reply services/frontend/pkg/config/config.go]: The `FRONTEND_OCS_RESOURCE_INFO_CACHE` vars have been adapted by now as well.

Regarding the auth information for redis those are to be given as part of the node address consistently now, e.g. `user:password@localhost:6789`.

**Comment by ownclouders** [comment]
:boom: Acceptance test [<strong>localApiTests-apiContract-ocis</strong>](https://drone.owncloud.com/owncloud/ocis/21491/50/1) failed. Further test are cancelled...

**Comment by mmattel** [comment]
Pls also change in `frontend/README.md` the envvar in text:

`FRONTEND_OCS_RESOURCE_INFO_CACHE_TYPE` --> `FRONTEND_OCS_RESOURCE_INFO_CACHE_STORE`

File: services/frontend/pkg/config/config.go [hunk 2]
```diff
@@ -123,27 +123,18 @@ type OCS struct {
...
-	ResourceInfoCacheTTL                 int                `yaml:"resource_info_cache_ttl" env:"FRONTEND_OCS_RESOURCE_INFO_CACHE_TTL" desc:"Max TTL in seconds for the resource info cache. 0 disables the cache."`
-	ResourceInfoCacheType                string             `yaml:"resource_info_cache_type" env:"FRONTEND_OCS_RESOURCE_INFO_CACHE_TYPE" desc:"The type of the resource info cache. Supported values are 'memory' and 'redis'."`
-	ResourceInfoCaches                   ResourceInfoCaches `yaml:"resource_info_caches,omitempty"` // only used for redis
+	ResourceInfoCacheType                string             `yaml:"resource_info_cache_type" env:"OCIS_CACHE_STORE;FRONTEND_OCS_RESOURCE_INFO_CACHE_STORE" desc:"The type of the resource info cache. Supported values are 'memory' and 'redis'."`
+	ResourceInfoCacheNodes               []string           `yaml:"resource_info_cache_nodes" env:"OCIS_CACHE_STORE_NODES;FRONTEND_OCS_RESOURCE_INFO_CACHE_STORE_NODES" desc:"The type of the resource info cache. Supported values are 'memory' and 'redis'."`
```
Comment (mmattel) [inline services/frontend/pkg/config/config.go:127]:
```suggestion
	ResourceInfoCacheNodes               []string           `yaml:"resource_info_cache_nodes" env:"OCIS_CACHE_STORE_NODES;FRONTEND_OCS_RESOURCE_INFO_CACHE_STORE_NODES" desc:"A comma separated list of nodes to access the configured store. This has no effect when 'memory' or 'ocmem' stores are configured. Note that the behaviour how nodes are used is dependent on the library of the configured store."`
```
Moste likely a copy/paste error 😅

File: services/frontend/pkg/config/config.go [hunk 3]
```diff
@@ -123,27 +123,18 @@ type OCS struct {
...
+	ResourceInfoCacheDatabase            string             `yaml:"resource_info_cache_database" env:"FRONTEND_OCS_RESOURCE_INFO_CACHE_DATABASE" desc:"The type of the resource info cache. Supported values are 'memory' and 'redis'."`
```
Comment (mmattel) [inline services/frontend/pkg/config/config.go:128]:
```suggestion
	ResourceInfoCacheDatabase            string             `yaml:"resource_info_cache_database" env:"FRONTEND_OCS_RESOURCE_INFO_CACHE_DATABASE" desc:"The database name the configured store should use."`
```
Most likely a copy/paste error 😅

File: services/gateway/pkg/config/config.go [hunk 4]
```diff
@@ -88,10 +89,13 @@ type StorageRegistry struct {
...
-	ProviderCacheTTL   int      `yaml:"provider_cache_ttl" env:"OCIS_CACHE_STORE_TTL;GATEWAY_PROVIDER_CACHE_TTL" desc:"Max TTL in seconds for the gateway's provider cache."`
-	CreateHomeCacheTTL int      `yaml:"create_home_cache_ttl" env:"OCIS_CACHE_STORE_TTL;GATEWAY_CREATE_HOME_CACHE_TTL" desc:"Max TTL in seconds for the gateway's create home cache."`
+	Store               string        `yaml:"store" env:"OCIS_CACHE_STORE;GATEWAY_CACHE_STORE" desc:"The type of the cache store. Supported values are: 'memory', 'ocmem', 'etcd', 'redis', 'redis-sentinel', 'nats-js', 'noop'. See the text description for details."`
+	Nodes               []string      `yaml:"nodes" env:"OCIS_CACHE_STORE_NODES;GATEWAY_CACHE_STORE_NODES" desc:"A comma separated list of nodes to access the configured store. This has no effect when 'memory' or 'ocmem' stores are configured. Note that the behaviour how nodes are used is dependent on the library of the configured store."`
+	Database            string        `yaml:"database" env:"GATEWAY_CACHE_DATABASE" desc:"Database name of the cache."`
```
Comment (mmattel) [inline services/gateway/pkg/config/config.go:94]:
```suggestion
	Database            string        `yaml:"database" env:"GATEWAY_CACHE_DATABASE" desc:"The database name the configured store should use."`
```
Harmonisation

**Comment by mmattel** [comment]
Q: why dont we have for the `xxx_CACHE_DATABASE` a global envvar?
We use that in all services but only have a local one.
Like
`GATEWAY_CACHE_DATABASE` -->
`OCIS_CACHE_DATABASE;GATEWAY_CACHE_DATABASE`

File: services/frontend/pkg/config/config.go [hunk 5]
```diff
@@ -123,27 +123,18 @@ type OCS struct {
...
 	AdditionalInfoAttribute              string             `yaml:"additional_info_attribute" env:"FRONTEND_OCS_ADDITIONAL_INFO_ATTRIBUTE" desc:"Additional information attribute for the user like {{.Mail}}."`
-	ResourceInfoCacheTTL                 int                `yaml:"resource_info_cache_ttl" env:"FRONTEND_OCS_RESOURCE_INFO_CACHE_TTL" desc:"Max TTL in seconds for the resource info cache. 0 disables the cache."`
-	ResourceInfoCacheType                string             `yaml:"resource_info_cache_type" env:"FRONTEND_OCS_RESOURCE_INFO_CACHE_TYPE" desc:"The type of the resource info cache. Supported values are 'memory' and 'redis'."`
-	ResourceInfoCaches                   ResourceInfoCaches `yaml:"resource_info_caches,omitempty"` // only used for redis
+	ResourceInfoCacheType                string             `yaml:"resource_info_cache_type" env:"OCIS_CACHE_STORE;FRONTEND_OCS_RESOURCE_INFO_CACHE_STORE" desc:"The type of the resource info cache. Supported values are 'memory' and 'redis'."`
```
Comment (mmattel) [inline services/frontend/pkg/config/config.go:126]: Are we sure we have a restriction to `redis` and `memory`?
`OCIS_CACHE_STORE` allows the full set...
Maybe to change the envvar description accordingly as we have this with other services.

File: services/frontend/pkg/config/config.go [hunk 6]
```diff
@@ -123,27 +123,18 @@ type OCS struct {
...
+	ResourceInfoCacheNodes               []string           `yaml:"resource_info_cache_nodes" env:"OCIS_CACHE_STORE_NODES;FRONTEND_OCS_RESOURCE_INFO_CACHE_STORE_NODES" desc:"The type of the resource info cache. Supported values are 'memory' and 'redis'."`
+	ResourceInfoCacheDatabase            string             `yaml:"resource_info_cache_database" env:"FRONTEND_OCS_RESOURCE_INFO_CACHE_DATABASE" desc:"The type of the resource info cache. Supported values are 'memory' and 'redis'."`
+	ResourceInfoCacheTable               string             `yaml:"resource_info_cache_table" env:"FRONTEND_OCS_RESOURCE_INFO_CACHE_TABLE" desc:"The type of the resource info cache. Supported values are 'memory' and 'redis'."`
```
Comment (mmattel) [inline services/frontend/pkg/config/config.go:129]:
```suggestion
	ResourceInfoCacheTable               string             `yaml:"resource_info_cache_table" env:"FRONTEND_OCS_RESOURCE_INFO_CACHE_TABLE" desc:"The database table the store should use."`
```
Moste likely a copy/paste error 😅

**Review by mmattel — APPROVED** [review]
LGTM from the docs pov

**Pull request merged**